File: hw/logic/mmu_v1_1.py
"""Plover v1.1 discrete MMU ??71024 PTE + IS62 main (sim / hwsim truth)."""

from __future__ import annotations

from dataclasses import dataclass, field

from hw.logic.cpld_decode import MemDecode, decode_addr

# Timing constants (ns) ??discrete-mmu-spec-v1.1.md
T_ACC_MMU_71024_NS = 15
T_ACC_MAIN_IS62_NS = 45
T_FAULT_COMB_NS = 5
T_MMU_PIPELINE_NS = T_ACC_MMU_71024_NS + T_ACC_MAIN_IS62_NS + T_FAULT_COMB_NS


@dataclass
class PteEntry:
    pa_hi: int  # PA[15:12], 4 bits
    valid: bool = True
    we: bool = True

    def pack(self) -> int:
        v = self.pa_hi & 0xF
        if self.valid:
            v |= 1 << 4
        if self.we:
            v |= 1 << 5
        return v & 0x3F

    @classmethod
    def unpack(cls, byte: int) -> PteEntry:
        b = byte & 0x3F
        return cls(
            pa_hi=b & 0xF,
            valid=bool(b & (1 << 4)),
            we=bool(b & (1 << 5)),
        )


def identity_pte_table() -> list[PteEntry]:
    """Boot handoff: VA page i maps to PA page i."""
    return [PteEntry(pa_hi=i, valid=True, we=True) for i in range(16)]


@dataclass
class Mmu71024State:
    """Behavioral 71024: 16 PTE bytes at A16=0; swap region not modeled in decode."""

    pte: list[PteEntry] = field(default_factory=identity_pte_table)

    def read_pte(self, va_page: int) -> PteEntry:
        return self.pte[va_page & 0xF]

    def write_pte(self, va_page: int, entry: PteEntry) -> None:
        self.pte[va_page & 0xF] = entry


@dataclass(frozen=True)
class MmuTranslateResult:
    va: int
    phys: int
    pte: PteEntry
    fault: bool
    nmi: bool
    mmu_bypass: bool
    mem: MemDecode


def _needs_mmu(va: int, map_mode: int, reset_active: bool) -> bool:
    """True when access is translated RAM (not ROM / mailbox / reset force)."""
    mem = decode_addr(va, map_mode, reset_active)
    if mem.mailbox_en or mem.addr_force_fffc:
        return False
    if mem.rom_cs_n == 0:
        return False
    if mem.ram1_cs_n == 0 or mem.ram2_cs_n == 0:
        return True
    return False


def fault_from_pte(pte: PteEntry, cpu_wr: bool) -> bool:
    if not pte.valid:
        return True
    if cpu_wr and not pte.we:
        return True
    return False


def translate_va(
    va: int,
    mmu: Mmu71024State,
    *,
    map_mode: int = 1,
    reset_active: bool = False,
    cpu_wr: bool = False,
) -> MmuTranslateResult:
    va &= 0xFFFF
    mem = decode_addr(va, map_mode, reset_active)
    bypass = not _needs_mmu(va, map_mode, reset_active)

    if bypass:
        return MmuTranslateResult(
            va=va,
            phys=va,
            pte=PteEntry(pa_hi=(va >> 12) & 0xF),
            fault=False,
            nmi=False,
            mmu_bypass=True,
            mem=mem,
        )

    pte = mmu.read_pte((va >> 12) & 0xF)
    fault = fault_from_pte(pte, cpu_wr)
    phys = ((pte.pa_hi & 0xF) << 12) | (va & 0xFFF)
    return MmuTranslateResult(
        va=va,
        phys=phys & 0xFFFF,
        pte=pte,
        fault=fault,
        nmi=fault,
        mmu_bypass=False,
        mem=decode_addr(phys, map_mode, reset_active),
    )


def decode_addr_v1_1(
    va: int,
    mmu: Mmu71024State,
    *,
    map_mode: int = 1,
    reset_active: bool = False,
    cpu_wr: bool = False,
) -> MmuTranslateResult:
    """Alias for translate_va ??v1.1 MMU decode entry point."""
    return translate_va(
        va, mmu, map_mode=map_mode, reset_active=reset_active, cpu_wr=cpu_wr
    )

File: crates/plover_mmu/src/mmu_v1_1.rs
//! Plover v1.1 discrete MMU ??71024 PTE table + IS62 main RAM.

use crate::decode::{DecodeResult, MapDecoder};

pub const T_ACC_MMU_71024_NS: u32 = 15;
pub const T_ACC_MAIN_IS62_NS: u32 = 45;
pub const T_FAULT_COMB_NS: u32 = 5;
pub const T_MMU_PIPELINE_NS: u32 = T_ACC_MMU_71024_NS + T_ACC_MAIN_IS62_NS + T_FAULT_COMB_NS;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct PteEntry {
    pub pa_hi: u8,
    pub valid: bool,
    pub we: bool,
}

impl PteEntry {
    pub fn pack(self) -> u8 {
        let mut v = self.pa_hi & 0xF;
        if self.valid {
            v |= 1 << 4;
        }
        if self.we {
            v |= 1 << 5;
        }
        v
    }

    pub fn unpack(byte: u8) -> Self {
        let b = byte & 0x3F;
        Self {
            pa_hi: b & 0xF,
            valid: (b & (1 << 4)) != 0,
            we: (b & (1 << 5)) != 0,
        }
    }
}

impl Default for PteEntry {
    fn default() -> Self {
        Self {
            pa_hi: 0,
            valid: true,
            we: true,
        }
    }
}

pub fn identity_pte_table() -> [PteEntry; 16] {
    std::array::from_fn(|i| PteEntry {
        pa_hi: i as u8,
        valid: true,
        we: true,
    })
}

#[derive(Clone, Debug)]
pub struct MmuV11 {
    pub pte: [PteEntry; 16],
}

impl Default for MmuV11 {
    fn default() -> Self {
        Self {
            pte: identity_pte_table(),
        }
    }
}

impl MmuV11 {
    pub fn read_pte(&self, va_page: u8) -> PteEntry {
        self.pte[(va_page & 0xF) as usize]
    }

    pub fn write_pte(&mut self, va_page: u8, entry: PteEntry) {
        self.pte[(va_page & 0xF) as usize] = entry;
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct TranslateResult {
    pub va: u16,
    pub phys: u16,
    pub pte: PteEntry,
    pub fault: bool,
    pub fault_nmi: bool,
    pub mmu_bypass: bool,
    pub mem: DecodeResult,
}

fn needs_mmu(dec: &DecodeResult) -> bool {
    !dec.mailbox
        && !dec.force_fffc
        && !dec.rom_cpu
        && (dec.ram1 || dec.ram2)
}

pub fn fault_from_pte(pte: PteEntry, cpu_wr: bool) -> bool {
    if !pte.valid {
        return true;
    }
    cpu_wr && !pte.we
}

pub fn translate(
    mmu: &MmuV11,
    decoder: &MapDecoder,
    va: u16,
    map_mode: u8,
    reset_active: bool,
    cpu_wr: bool,
) -> TranslateResult {
    let mem = decoder.decode(va, map_mode, reset_active);
    if !needs_mmu(&mem) {
        return TranslateResult {
            va,
            phys: va,
            pte: PteEntry {
                pa_hi: ((va >> 12) & 0xF) as u8,
                valid: true,
                we: true,
            },
            fault: false,
            fault_nmi: false,
            mmu_bypass: true,
            mem,
        };
    }

    let pte = mmu.read_pte(((va >> 12) & 0xF) as u8);
    let fault = fault_from_pte(pte, cpu_wr);
    let phys = u16::from(pte.pa_hi & 0xF) << 12 | (va & 0x0FFF);
    let mem_phys = decoder.decode(phys, map_mode, reset_active);

    TranslateResult {
        va,
        phys,
        pte,
        fault,
        fault_nmi: fault,
        mmu_bypass: false,
        mem: mem_phys,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn identity_map_roundtrip() {
        let mmu = MmuV11::default();
        let dec = MapDecoder;
        let r = translate(&mmu, &dec, 0x1234, 1, false, false);
        assert!(!r.fault);
        assert_eq!(r.phys, 0x1234);
    }

    #[test]
    fn invalid_pte_faults() {
        let mut mmu = MmuV11::default();
        mmu.write_pte(
            4,
            PteEntry {
                pa_hi: 4,
                valid: false,
                we: true,
            },
        );
        let dec = MapDecoder;
        let r = translate(&mmu, &dec, 0x4000, 1, false, false);
        assert!(r.fault_nmi);
    }
}
